File: 0149-max-points-on-a-line/0149-max-points-on-a-line.js
function maxPoints(points) {
    const numPoints = points.length;

    // 최소 1개의 점은 항상 존재
    let maxPointsOnLine = 1;

    // 각 점을 첫 번째 점으로 설정
    for (let i = 0; i < numPoints; i++) {
        const x1 = points[i][0];
        const y1 = points[i][1];

        // 두 번째 점을 선택하여 직선 생성
        for (let j = i + 1; j < numPoints; j++) {
            const x2 = points[j][0];
            const y2 = points[j][1];

            // i, j 두 점은 이미 같은 직선 위에 있음
            let pointsOnCurrentLine = 2;

            // 나머지 점들이 같은 직선 위에 있는지 확인
            for (let k = j + 1; k < numPoints; k++) {
                const x3 = points[k][0];
                const y3 = points[k][1];

                // 외적을 이용한 일직선 여부 확인
                //
                // (y2 - y1) / (x2 - x1)
                // =
                // (y3 - y1) / (x3 - x1)
                //
                // 나눗셈을 피하기 위해 교차 곱
                if (
                    (y2 - y1) * (x3 - x1) ===
                    (y3 - y1) * (x2 - x1)
                ) {
                    pointsOnCurrentLine++;
                }
            }

            // 최댓값 갱신
            maxPointsOnLine = Math.max(
                maxPointsOnLine,
                pointsOnCurrentLine
            );
        }
    }

    return maxPointsOnLine;
}

File: leetcode/0149-max-points-on-a-line/0149-max-points-on-a-line.java
class Solution {
    public int maxPoints(int[][] points) {
        int n = points.length;
        int maxPointsOnLine = 1; // At least one point exists
      
        // Iterate through each point as the first point of a potential line
        for (int i = 0; i < n; ++i) {
            int x1 = points[i][0];
            int y1 = points[i][1];
          
            // Iterate through each point after i as the second point to form a line
            for (int j = i + 1; j < n; ++j) {
                int x2 = points[j][0];
                int y2 = points[j][1];
                int pointsOnCurrentLine = 2; // Line through points i and j has at least 2 points
              
                // Check all remaining points to see if they lie on the same line
                for (int k = j + 1; k < n; ++k) {
                    int x3 = points[k][0];
                    int y3 = points[k][1];
                  
                    // Check if point k is collinear with points i and j
                    // Using cross product: if (y2-y1)/(x2-x1) = (y3-y1)/(x3-x1)
                    // Cross multiply to avoid division: (y2-y1)*(x3-x1) = (y3-y1)*(x2-x1)
                    int crossProduct1 = (y2 - y1) * (x3 - x1);
                    int crossProduct2 = (y3 - y1) * (x2 - x1);
                  
                    if (crossProduct1 == crossProduct2) {
                        ++pointsOnCurrentLine;
                    }
                }
              
                // Update the maximum number of points found on a single line
                maxPointsOnLine = Math.max(maxPointsOnLine, pointsOnCurrentLine);
            }
        }
      
        return maxPointsOnLine;
    }
}
